plugins/generic/shibboleth/pages/ShibbolethHandler.inc.php

```diff
@@ -0,0 +1,437 @@
+			// We use the e-mail as a key.
+			if (empty($userEmail)) {
+				syslog(LOG_ERR, "Shibboleth failed to find UIN $uin and no email given.");
+				Validation::logout();
+				Validation::redirectLogin();
+				return false;
+			}
+			$user =& $userDao->getUserByEmail($userEmail);
+
+			if (isset($user)) {
+				syslog(LOG_INFO, "Shibboleth located returning email $userEmail");
+
+				if ($user->getAuthStr() != "") {
+					syslog(
+						LOG_ERR,
+						"Shibboleth user with email $userEmail already has UID"
+					);
+					Validation::logout();
+					Validation::redirectLogin();
+					return false;
+				} else {
+					$user->setAuthStr($uin);
+					$userDao->updateObject($user);
+				}
+			} else {
+				$user = $this->_registerFromShibboleth();
+			}
+		}
+
+		if (isset($user)) {
+			$this->_checkAdminStatus($user);
+
+			$disabledReason = null;
+			$success = Validation::registerUserSession($user, $disabledReason);
+
+			if (!$success) {
+				// @@@ TODO: present user with disabled reason
+				syslog(
+					LOG_ERR,
+					"Disabled user $uin attempted Shibboleth login" .
+						(is_null($disabledReason) ? "" : ": $disabledReason")
+				);
+				Validation::logout();
+				Validation::redirectLogin();
+				return false;
+			}
+
+			return $this->_redirectAfterLogin($request);
+		}
+
+		return false;
+	}
+
+	/**
+	 * @copydoc ShibbolethHandler::activateUser()
+	 */
+	function signIn($args, $request) {
+		return $this->_shibbolethRedirect($args, $request);
+	}
+
+	/**
+	 * Intercept normal logout; redirect to context home page instead
+	 * of login (which would send back to Shibboleth again).
+	 * 
+	 * @param $args array
+	 * @param $request Request
+	 * @return bool
+	 */
+	function signOut($args, $request) {
+		$context = $this->getTargetContext($request);
+		$router = $request->getRouter();
+
+		Validation::logout();
+
+		$contextPath = is_null($context) ? "" : $context->getPath();
+		$returnUrl = $router->url($request, $contextPath);
+		return $request->redirectUrl($returnUrl);
+	}
+
+	/**
+	 * @copydoc ShibbolethHandler::activateUser()
+	 */
+	function validate($args, $request) {
+		return $this->_shibbolethRedirect($args, $request);
+	}
+
+
+	//
+	// Private helper methods
+	//
+	/**
+	 * Get the Shibboleth plugin object
+	 * 
+	 * @return ShibbolethAuthPlugin
+	 */
+	function &_getPlugin() {
+		$plugin =& PluginRegistry::getPlugin('generic', SHIBBOLETH_PLUGIN_NAME);
+		return $plugin;
+	}
+
+	/**
+	 * Check if the user should be an admin according to the
+	 * Shibboleth plugin settings, and adjust the User object
+	 * accordingly.
+	 * 
+	 * @param $user User
+	 */
+	function _checkAdminStatus($user) {
+		$adminsStr = $this->_plugin->getSetting(
+			$this->_contextId,
+			'shibbolethAdminUins'
+		);
+		$admins = explode(' ', $adminsStr);
+
+		$uin = $user->getAuthStr();
+		if (empty($uin)) {
+			return;
+		}
+
+		$userId = $user->getId();
+		$adminFound = array_search($uin, $admins);
+
+		$userGroupDao =& DAORegistry::getDAO('UserGroupDAO');
```

Comment:
Unnecessary reference use (probably also elsewhere) -- there are lots of these in the codebase, but we're trying to clean them up as we go.

Reply:
More cut-and-paste.